odra-core/src/host_env.rs

```diff
@@ -131,30 +131,85 @@ impl HostEnv {
         let backend = self.backend.borrow();
 
         let bytes = backend.get_event(contract_address, index)?;
-        let event_name = Self::extract_event_name(&bytes)?;
-        if event_name == format!("event_{}", T::name()) {
-            T::from_bytes(&bytes)
-                .map_err(|_| EventError::Parsing)
-                .map(|r| r.0)
-        } else {
-            Err(EventError::UnexpectedType(event_name))
+        T::from_bytes(&bytes)
+            .map_err(|_| EventError::Parsing)
+            .map(|r| r.0)
+    }
+
+    pub fn get_event_bytes(
+        &self,
+        contract_address: &Address,
+        index: i32
+    ) -> Result<Bytes, EventError> {
+        let backend = self.backend.borrow();
+        backend.get_event(contract_address, index)
+    }
+
+    pub fn contract_event_names(&self, contract_address: &Address) -> Vec<String> {
+        let backend = self.backend.borrow();
+        let events_count = backend.get_events_count(contract_address);
+        let mut event_names = vec![];
+        for event_id in 0..events_count {
+            let bytes = backend
+                .get_event(contract_address, event_id as i32)
+                .unwrap();
+            let event_name = Self::extract_event_name(&bytes).unwrap();
+            event_names.push(event_name);
         }
+        event_names
     }
 
-    pub fn get_events_count(&self, contract_address: &Address) -> u32 {
+    pub fn events(&self) -> Vec<Bytes> {
+        let mut events = vec![];
+        self.events_count
+            .borrow()
+            .iter()
+            .for_each(|(contract_address, count)| {
+                for i in 0..*count {
+                    events.push(self.get_event_bytes(contract_address, i as i32).unwrap());
+                }
+            });
+        events
+    }
+
+    pub fn event_names(&self) -> Vec<String> {
+        let mut event_names = vec![];
+        self.events_count
+            .borrow()
+            .iter()
+            .for_each(|(contract_address, _)| {
+                event_names.append(&mut self.contract_event_names(contract_address));
+            });
+        event_names
```

Comment:
Could be written without a tmp vec
```suggestion
        self.events_count
            .borrow()
            .iter()
            .map(|(contract_address, _)| self.contract_event_names(contract_address))
            .flatten() // to flateen Vec<Vec<String>>
            .collect()
```

Reply:
Done